docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+- The node registration 
+
+
+### Non-Goals/Future Work
+
+- Not to change assumptions around management cluster to workload cluster connectivity
+- Not to solve the protection of initial cluster bootstrap secrets for the control plane nodes (see Node Bootstrapper)
+
+## Proposal
+
+
+
+| Component | Location | Description |
+| --------- | -------- | ----------- |
+| node-attestation-controller     | Part of Cluster API repository, under bootstrap/node/attestation, and imported by Cluster API infra providers for implementation. A generic challenge-response implementation will be included for providers / bare metal without an attestation mechanism.     |   A controller to verify and sign the CSR. this would be typically an importable controller where the infrastructure provider implements the interface with specific code for CSR approval + Start the controller as part of its main.go or through an independent binary.   |
+
+#### Kubelet authentication plugin
+
+![sequence diagram](images/kubelet-authentication/bootstrap-flow.png)
+
+
+
+The following sequence diagram shows at a high level how a nodeConfig enables bootstrapping a kubernetes node. This sequence diagram is independent of API definition and alternatives and is relevant in all cases.
+
+We propose a kubelet authentication plugin to be present on the instances, (how the nodeadm CLI made available is out of scope for this proposal), which will be responsible for node registration, as well as certificate rotation. The agent will be made up of two parts:
+- A common library vendored from Cluster API which includes the following functionality:
+    - Certificate filesystem locking
+    - Checking existing certificate validity
+    - Certificate signing request generation for both kubernetes client and serving certificates
+    - Submission of CSRs to the API server and waiting for approval
+- A provider specific implementation for node attestation
+    - A provider will only need to implement the generation of the attestation to be included in the CSR and the retrieval of the provider ID to be stored in an X.509 extension attribute.
+
+The behaviour of the authentication plugin will be as follows:
+
+
+#### Node Attestation
+
+As for the node-attestation-controller, the following interface needs to be implemented by the infrastructure providers:
+```go=
+type ClusterAPISigner interface {
+     VerifyClientAttestationData (csr *certificatesv1beta1.CertificateSigningRequest) err
+     VerifyServingAttestationData (csr *certificatesv1beta1.CertificateSigningRequest) err
+}
+```
+
+
+This will enable infrastructure providers to perform infrastructure-specific validation of node attestations (TPM, appended tags by the provider etc…)
+
+Cluster API will be responsible for partially verifying node identity through verifying that:
+
+- a corresponding machine object exist for the `.spec.Username` ( `system:nodes:<nodename>`)
+- The Machine state is provisioning
```

Comment:
We shouldn't rely on the state to determine a state, but rather look at all the other variables to determine that the machine hasn't been provisioned, for example Status.NodeRef is not present.

Reply:
ack